### ISSUE TITLE
Add Decimal type

### ISSUE THREAD

**Issue by mnordine** [dart-lang/language#889]
TC39 Proposal: https://github.com/tc39/proposal-decimal

**Comment by lrhn** [comment]
A Dart decimal type as a language addition would probably be: 
* A type extending `num`, separate from `int` and double`.
* Allow literals, whetever syntax seems reasonable (`124.56d`, `0d124.56`, `123.45e-5d` or `123.45d-5` or neither?`).
* Likely allow adding or being added to `int`s or `double`s. It's unclear what type the result is, but likely `decimal + int` is `decimal` and `decimal + double` is `double`.



Arguably, this can be either a library issue or a language issue depending on how deep the intgration goes. Let's keep it as a language issue.

**Comment by mnordine** [comment]
Sorry, I mistakenly put it in sdk repo.

**Comment by mnordine** [comment]
Also want to mention the issue for `BigInt` suffix here: https://github.com/dart-lang/language/issues/372

**Comment by cedvdb** [comment]
> * `decimal + double` is `double`

why ? If one uses decimal in the first place, chances are the next calculation is also supposed to be

**Comment by mateusfccp** [comment]
> > * `decimal + double` is `double`
> 
> why ? If one uses decimal in the first place, chances are the next calculation is also supposed to be

Because adding a double to a Decimal can't always keep the arbitrary precision.

**Comment by cedvdb** [comment]
@mateusfccp 

Genuine question, do you have an example ? Using the decimal package:

```
 Decimal.parse(0.3.toString()) + Decimal.parse(0.6.toString())
```

Gives 0.9 for 0.3 + 0.6

**Comment by mateusfccp** [comment]
> @mateusfccp 
> 
> Genuine question, do you have an example ? Using the decimal package:
> 
> ```
>  // assuming dart would internally do something similar
>  Decimal.parse(0.3.toString()) + Decimal.parse('0.6')
> ```
> 
> Gives 0.9 for 0.3 + 0.6.

Sure, if you assume that Dart would implicitly convert double to decimal, then it could work.

I don't know if it is a good idea, though, because doing so will potentially result in information being lost.

It's similar to `1 + 1.5`. Instead of converting `1.5` to an int (what would it be? Truncated? Rounded?) we convert `1` to `1.0` and the result is a double.

**Comment by cedvdb** [comment]
> I don't know if it is a good idea, though, because doing so will potentially result in information being lost.

Do you have an example ?

**Comment by tatumizer** [comment]
> Do you have a concrete example ?  

```dart
print(Decimal.parse('0.6').toDouble().toStringAsFixed(20));  // prints 0.59999999999999997780
```

**Comment by cedvdb** [comment]
> ```dart
> .toStringAsFixed(20)
> ```

That's a counter example 


```dart
print(0.6.toStringAsFixed(20));  // prints 0.59999999999999997780

```

**Comment by tatumizer** [comment]
Sorry, I posted by mistake. The loss of precision occurs whenever you convert from Decimal to double or vice versa. You are probably fine if you keep everything in Decimal, except that performance would suffer.

**Comment by cedvdb** [comment]
> The loss of precision occurs whenever you convert from  Decimal to double or vice versa

Sure, the more bases you convert a number to, the more likely you are to hit a recurring suite of digits for a base, but that doesn't explain why double over decimal.   

> except that performance would suffer.

Usage and expectations have to be taken into account too.

**Comment by tatumizer** [comment]
`Decimal` is implemented as a thin wrapper around Rational, which stores the numerator and denominator *exactly*, as BigInts, the denominator being a power of 10. They dynamically allocate as much memory as needed.

**Comment by cedvdb** [comment]
Nice, note that the TC39 proposal uses Decimal128 though

**Comment by lrhn** [comment]
Indeed. If Dart adds a `Decimal` type, it will very likely match what ECMAScript does, so that it's efficiently compilable to the web. 
If that type would be `Decimal128` that is not directly compatible with `double`, then Dart would probably not make it implement `num`. If ECMAScript allows a decimal and a double to be added, then Dart might make `Decimal` implement `num`, and have the same behavior as doing the same thing in ECMAScript.

So far, if I quick-read the summaries correctly, the expected choice is to not allow `double + decimal`, and likely not `decimal + double`, so the Dart type would also not be a `num`. It's a separate numeric type, like `BigInt`, that you have to explicitly inject and project values from.
